File: tools/commands/ci/github/verify-glibc-symbols.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//! ```

use std::{
    collections::BTreeSet,
    env,
    path::Path,
    process::{Command, ExitCode, Stdio},
};

const DEFAULT_MAX_GLIBC: &str = "2.36";

#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
struct Version {
    major: u64,
    minor: u64,
    patch: u64,
}

impl Version {
    fn text(&self) -> String {
        if self.patch == 0 {
            format!("{}.{}", self.major, self.minor)
        } else {
            format!("{}.{}.{}", self.major, self.minor, self.patch)
        }
    }
}

fn main() -> ExitCode {
    match verify() {
        Ok(()) => ExitCode::SUCCESS,
        Err(error) => {
            eprintln!("{error}");
            ExitCode::from(1)
        }
    }
}

fn verify() -> Result<(), String> {
    let (max, files) = parse_args(env::args().skip(1))?;
    if files.is_empty() {
        return Err(
            "Usage: rust-script tools/commands/ci/github/verify-glibc-symbols.rs [--max 2.36] <file.node>..."
                .to_string(),
        );
    }

    let mut failed = false;
    for file in files {
        if !Path::new(&file).exists() {
            report_error(&format!("native binary does not exist: {file}"));
            failed = true;
            continue;
        }

        let versions = parse_versions(&readelf_version_info(&file)?);
        if versions.is_empty() {
            report_error(&format!(
                "native binary has no GLIBC_* version records: {file}"
            ));
            failed = true;
            continue;
        }
        if versions.iter().any(|version| version > &max) {
            let highest = versions.last().expect("nonempty versions");
            report_error(&format!(
                "{file} requires GLIBC_{}, above the supported ceiling GLIBC_{}",
                highest.text(),
                max.text()
            ));
            failed = true;
            continue;
        }
        let highest = versions.last().expect("nonempty versions");
        println!("{file}: GLIBC_{} <= GLIBC_{}", highest.text(), max.text());
    }

    if failed {
        Err("glibc symbol verification failed".to_string())
    } else {
        Ok(())
    }
}

fn parse_args(args: impl Iterator<Item = String>) -> Result<(Version, Vec<String>), String> {
    let mut max = DEFAULT_MAX_GLIBC.to_string();
    let mut files = Vec::new();
    let mut args = args.peekable();
    while let Some(arg) = args.next() {
        if arg == "--max" {
            max = args
                .next()
                .ok_or_else(|| "--max requires a value".to_string())?;
        } else {
            files.push(arg);
        }
    }
    Ok((parse_version(&max)?, files))
}

fn readelf_version_info(file: &str) -> Result<String, String> {
    let output = Command::new("readelf")
        .args(["--version-info", file])
        .stdin(Stdio::null())
        .output()
        .map_err(|error| format!("failed to run readelf: {error}"))?;
    if !output.status.success() {
        return Err(format!(
            "readelf --version-info failed for {file}\n{}{}",
            String::from_utf8_lossy(&output.stderr),
            String::from_utf8_lossy(&output.stdout)
        )
        .trim()
        .to_string());
    }
    Ok(format!(
        "{}\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    ))
}

fn parse_versions(text: &str) -> BTreeSet<Version> {
    let mut versions = BTreeSet::new();
    let mut rest = text;
    while let Some(index) = rest.find("GLIBC_") {
        rest = &rest[index + "GLIBC_".len()..];
        let end = rest
            .find(|character: char| !character.is_ascii_digit() && character != '.')
            .unwrap_or(rest.len());
        if let Ok(version) = parse_version(&rest[..end]) {
            versions.insert(version);
        }
        rest = &rest[end..];
    }
    versions
}

fn parse_version(value: &str) -> Result<Version, String> {
    let parts = value.split('.').collect::<Vec<_>>();
    let ([major, minor] | [major, minor, _]) = parts.as_slice() else {
        return Err(format!(
            "glibc version must look like MAJOR.MINOR[.PATCH], got {value}"
        ));
    };
    Ok(Version {
        major: major.parse().map_err(|_| {
            format!("glibc version must look like MAJOR.MINOR[.PATCH], got {value}")
        })?,
        minor: minor.parse().map_err(|_| {
            format!("glibc version must look like MAJOR.MINOR[.PATCH], got {value}")
        })?,
        patch: parts
            .get(2)
            .map(|patch| patch.parse())
            .transpose()
            .map_err(|_| format!("glibc version must look like MAJOR.MINOR[.PATCH], got {value}"))?
            .unwrap_or(0),
    })
}

fn report_error(message: &str) {
    eprintln!("::error::{message}");
}

File: tools/support/compat/github/verify-glibc-symbols.mjs
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import { pathToFileURL } from "node:url";

const DEFAULT_MAX_GLIBC = "2.36";

export function parseGlibcVersions(text) {
  const versions = new Map();
  for (const match of text.matchAll(/\bGLIBC_(\d+)\.(\d+)(?:\.(\d+))?\b/g)) {
    const patch = match[3] == null ? 0 : Number(match[3]);
    const version = {
      major: Number(match[1]),
      minor: Number(match[2]),
      patch,
      text: `${Number(match[1])}.${Number(match[2])}${patch === 0 ? "" : `.${patch}`}`,
    };
    versions.set(version.text, version);
  }
  return [...versions.values()].sort(compareGlibcVersions);
}

export function parseGlibcVersion(value) {
  const match = /^(\d+)\.(\d+)(?:\.(\d+))?$/.exec(value);
  if (!match) {
    throw new Error(`glibc version must look like MAJOR.MINOR[.PATCH], got ${value}`);
  }
  const patch = match[3] == null ? 0 : Number(match[3]);
  return {
    major: Number(match[1]),
    minor: Number(match[2]),
    patch,
    text: `${Number(match[1])}.${Number(match[2])}${patch === 0 ? "" : `.${patch}`}`,
  };
}

export function compareGlibcVersions(left, right) {
  return left.major - right.major || left.minor - right.minor || left.patch - right.patch;
}

export function highestGlibcVersion(versions) {
  return versions.at(-1) ?? null;
}

export function glibcVersionsAbove(versions, maxVersion) {
  return versions.filter((version) => compareGlibcVersions(version, maxVersion) > 0);
}

function readElfVersionInfo(filePath) {
  const result = spawnSync("readelf", ["--version-info", filePath], {
    encoding: "utf8",
  });
  if (result.error != null) {
    throw result.error;
  }
  if (result.status !== 0) {
    throw new Error(
      `readelf --version-info failed for ${filePath}\n${result.stderr}${result.stdout}`.trim(),
    );
  }
  return `${result.stdout}\n${result.stderr}`;
}

function parseArgs(args) {
  let max = DEFAULT_MAX_GLIBC;
  const files = [];
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (arg === "--max") {
      const value = args[index + 1];
      if (!value) {
        throw new Error("--max requires a value");
      }
      max = value;
      index += 1;
      continue;
    }
    files.push(arg);
  }
  return { files, max: parseGlibcVersion(max) };
}

function reportError(message) {
  console.error(`::error::${message}`);
}

function main() {
  const { files, max } = parseArgs(process.argv.slice(2));
  if (files.length === 0) {
    throw new Error(
      "Usage: rust-script tools/commands/ci/github/verify-glibc-symbols.rs [--max 2.36] <file.node>...",
    );
  }

  let failed = false;
  for (const file of files) {
    if (!fs.existsSync(file)) {
      reportError(`native binary does not exist: ${file}`);
      failed = true;
      continue;
    }

    const versions = parseGlibcVersions(readElfVersionInfo(file));
    if (versions.length === 0) {
      reportError(`native binary has no GLIBC_* version records: ${file}`);
      failed = true;
      continue;
    }

    const tooNew = glibcVersionsAbove(versions, max);
    if (tooNew.length > 0) {
      const highest = highestGlibcVersion(versions);
      reportError(
        `${file} requires GLIBC_${highest.text}, above the supported ceiling GLIBC_${max.text}`,
      );
      failed = true;
      continue;
    }

    const highest = highestGlibcVersion(versions);
    console.log(`${file}: GLIBC_${highest.text} <= GLIBC_${max.text}`);
  }

  if (failed) {
    process.exitCode = 1;
  }
}

if (process.argv[1] != null && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main();
}
